sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TimestampFormatter.scala

```diff
@@ -163,27 +165,66 @@ class Iso8601TimestampFormatter(
   protected lazy val legacyFormatter = TimestampFormatter.getLegacyFormatter(
     pattern, zoneId, locale, legacyFormat)
 
+  override def parseOptional(s: String): Option[Long] = {
+    try {
+      val parsed = formatter.parseUnresolved(s, new ParsePosition(0))
```

Comment:
It will be check in https://github.com/apache/spark/pull/41078/files/77c78b1f0ed3f7ee3c301eec2fa543a048c35531#diff-b42bcba727feeebf78f0e5540f2d4f6c6a38afd2225e4ebeae22a604e42eb094R183-R188
. The logic can't go wrong. But in this situation, the speed up will not work.

Reply:
> Could you re-check the example from the docs: '2012-00-65' and add a test, please.

Done